lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -289,6 +294,106 @@ def filter_query(filters: list[str], query: Query) -> Query:
 
         return query.filter(and_(*and_list))
 
+    def accumulate(self, keys: JSONOBJECT, key: str, value: Any):
+        """Recursive helper to accumulate the metadata namespace
+
+        Iterate through a list of metadata key/value pairs to construct a
+        herarchical aggregation of all metadata keys across the selected
+        datasets. Each key in the hierarchy is represented as a key in a
+        nested JSON object. "Leaf" keys are represented with empty JSON
+        values. E.g.,
+
+            {
+                "dataset": {"name": {}, "metalog": {"pbench": {"script": {}}}},
+                "server": {"deletion": {}, "tarball-path": {}},
+                "global": {"server": {"legacy": {"sha1": {}}}}
+            }
+
+        Args:
+            keys: a JSONOBJECT to update with the recursive key/value
+            key: the current metadata key path element
+            value: the current metadata key's value
+        """
+        if key in keys:
+            p = keys[key]
+        else:
+            p = {}
+            keys[key] = p
+        if isinstance(value, dict):
+            for k, v in value.items():
+                self.accumulate(p, k, v)
+
+    def keyspace(self, query: Query) -> JSONOBJECT:
+        """Aggregate the dataset metadata keyspace
+
+        Run the query we've compiled, but instead of returning Dataset proxies,
+        we only want the metadata key/value pairs we've selected.
+
+        NOTE: In general we expect every dataset to have some metadata; however
+        in the event one doesn't, the key and value returned by the joined
+        query will be None. We handle this by ignoring those values.
```

Comment:
I don't see how this NOTE applies to the code in this function...what am I missing?  It looks like this function always returns at least `{ "dataset": { <Dataset.__table__.columns> } }`.  Is this note saying that the call to `query...all()` might return `None`?  If so, then, rather than saying that we "ignore" something, it would be better to say that we just return the "seed" values.

Reply:
Noted: I'll make it a bit more explicit. In particular, this refers to the `elif` inside the loop: and in the current structure of the code that connection is apparently a bit too subtle. It's a SQL [LEFT] OUTER JOIN thing.

Specifically, `if not k` then we're processing a returned row for a dataset that didn't have any metadata at all, and we silently skip it. The left outer join returns a row for each row in the left table (datasets); and if there's a foreign key relationship that doesn't match on the right, the "missing" right columns are filled with nulls. (The default inner join wouldn't return these rows at all, which was my basic problem when I coded `filter` ... it meant the datasets list API couldn't see datasets without metadata.) In this particular case, I've mangled the `SELECT` to return only the right table columns, so we get two nulls.

As the comment points out, this is rare and even improbable for a real server, but definitely happens in our unit test fixtures.

Reply:
> this refers to the elif inside the loop

Ah, OK.  (Perhaps it would be better if the NOTE were positioned closer to the code that it's describing?)

So, `query...all()` might return a list with an entry which is a pair in which the first of the pair is false-y?  That's an interesting implementation choice on SQL(Alchemy)'s part.  (I think I might have omitted that entry from the list....)

Reply:
You can use triple-double-quotes to do block comments.  Normally, we only do that for docstrings, but we _could_ use them elsewhere.  (Personally, I don't have much problem with block comments made out of line comments, and I'd rather save triple-double-quotes for docstrings and other "special strings", but if we need 'em they are there.)